crates/autopilot/src/run_loop.rs

```diff
@@ -389,6 +399,83 @@ impl RunLoop {
         .collect()
     }
 
+    /// Returns true if winning solution is fair or winner is None
+    fn check_fairness(
+        winner: Option<&Participant>,
+        remaining: &Vec<Participant>,
+        auction: &domain::Auction,
+    ) -> bool {
+        let winner = match winner {
+            Some(winner) => winner,
+            None => return true,
+        };
+        let fairness_threshold = match winner.driver.fairness_threshold {
+            Some(threshold) => threshold,
+            None => return true,
+        };
+
+        // Returns Some(buy token amount) if left is better than right.
+        let improvement_in_buy = |left: &TradedAmounts, right: &TradedAmounts| {
```

Comment:
> might be better as a separate function

Really? It's only used in the context of this very specialised thing and pretty small in and of itself, making it a top level function on the module just moves it further away from the place where it needs to be understood, but if you feel storngly,

>it's unclear to me when using fallback values is desired and when shortcutting with ?

Replaced everything with `?` as I guess overflows for executed amounts aren't really a thing (likely solutions that overflow don't even make it this far).

> does this work correctly when 2 execution fill the same order by different amounts?

It should (I'm comparing exchange rates which for different partial fills should still find the better execution).

Implemented 2 & 3, let me know if this is easier to read now.

Reply:
Wasn't sure about `1` when I suggested it either.

Not sure if it got overlooked in my other blurb but by operating on the underlying `U256` and using `.full_mul()` which prevents overflow this could become a lot less code. To a point where I think it's small enough to not warrant a separate function.

```rust
            let right_sell_left_buy = right.sell.0.full_mul(left.buy.0);
            let left_sell_right_buy = left.sell.0.full_mul(right.buy.0);
            let improvement = right_sell_left_buy.saturating_sub(left_sell_right_buy);

            // The difference divided by the original sell amount is the improvement in buy
            // token.
            improvement
                .checked_div(right.sell.0.into())
                .and_then(|div| U256::try_from(div).ok())
                .unwrap_or_default()
```